### PR TITLE
[11.0][ADD] Scrap reason code module

### PR DESCRIPTION
**Pull request by bodedra**
Usage
--------

In the Inventory module, open the Configuration menu and select Scrap Reason Codes.
Create the required scrap reason codes. Under Operations, select Scrap. Click the 
create button to create a new scrap order. You will see a reason code field on the
scrap form which will allow you to select any of the scrap codes you created previously.

### CODE REVIEW

**Comment by bodedra** [comment]
@hveficent @aheficent @lreficent @jbeficent @max3903 

Would you please help me.

Test case is failed in `stock_request` or `stock_request_analytic` module.

Firefox_Screenshot_2019-11-27T08-06-10 411Z

**Comment by bodedra** [comment]
Are you available for review?

@b-kannan @osimallen  @smangukiya @gavindav @diverfr @mdodoo @guimarc-br

Test case are green for this PR.

Firefox_Screenshot_2019-11-27T08-03-04 972Z

**Comment by bodedra** [comment]
ci is :green_apple:

**Comment by HviorForgeFlow** [comment]
@bodedra all I can see is that stock_request_user is not available to read stock.warehouse0 record.
Can't see why

**Comment by bodedra** [comment]
> @bodedra all I can see is that stock_request_user is not available to read stock.warehouse0 record.
> Can't see why

@hveficent Agree. I verified that read access right is already set. Here is ref https://github.com/OCA/stock-logistics-warehouse/blob/11.0/stock_request/security/ir.model.access.csv#L14

**Comment by bodedra** [comment]
@rousseldenis Is there any suggestion for fix https://github.com/OCA/stock-logistics-warehouse/pull/770#issuecomment-558977212

Read access right already for warehouse.  https://github.com/OCA/stock-logistics-warehouse/blob/11.0/stock_request/security/ir.model.access.csv#L14

**Comment by HviorForgeFlow** [comment]
@bodedra 
Ugly form to fix error:
```python            
if self.warehouse_id != loc_wh:
    self.location_id = self.warehouse_id.sudo().lot_stock_id.id
if self.warehouse_id.sudo().company_id != self.company_id:
    self.company_id = self.warehouse_id.sudo().company_id
```

**Comment by HviorForgeFlow** [comment]
@etobella can you figure out how to avoid sudo usage here? I think as you only writing/reading fields it does not matters.

**Comment by bodedra** [comment]
This PR is :green_apple: and ready for your review.

@b-kannan @dreispt @max3903  @gavindav @diverfr @mdodoo @guimarc-br

**Comment by bodedra** [comment]
Thank you @hveficent for suggested fix in stock_request_* module.

File: scrap_reason_code/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+# Copyright (C) 2019 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'Scrap Reason Code',
+    'version': '11.0.1.0.0',
+    'license': 'AGPL-3',
+    'summary': 'Reason code for scrapping',
+    'author': 'Open Source Integrators, Odoo Community Association (OCA)',
+    'website': 'http://www.opensourceintegrators.com',
```
Comment (nikul-serpentcs) [inline scrap_reason_code/__manifest__.py]: @bodedra URL should be https://github.com/OCA/stock-logistics-warehouse

**Review by nikul-serpentcs — COMMENTED** [review]
nitpicking

Squash Commit's

**Comment by bodedra** [comment]
@nikul-serpentcs Comment addressed! Thanks for your review.

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1:

File: scrap_reason_code/models/reason_code.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Copyright (C) 2019 IBM Corp.
+# Copyright (C) 2019 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ReasonCode(models.Model):
+    _name = 'reason.code'
```
Comment (rousseldenis) [inline scrap_reason_code/models/reason_code.py]: IMHO either this is a generic model and should go into a dedicated module, either you should be more specific e.g.: 'scrap.reason.code'

Comment (bodedra) [reply scrap_reason_code/models/reason_code.py]: Done.

**Comment by bodedra** [comment]
@rousseldenis Comment attended!

**Comment by bodedra** [comment]
@Chandresh-SerpentCS are you available for a review?

File: scrap_reason_code/README.rst [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+**Table of contents**
+
+.. contents::
+   :local:
+
+Usage
+=====
+
+In the Inventory module, open the Configuration menu and select Scrap Reason Codes.
```
Comment (rousseldenis) [inline scrap_reason_code/README.rst:39]: You need to improve this and include the behaviour of selecting code and scrap location automatically filled in.
A little bit more explanation of use cases would be welcome.

File: scrap_reason_code/models/stock_scrap.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+class StockScrap(models.Model):
+    _inherit = 'stock.scrap'
+
+    reason_code_id = fields.Many2one(
+        'scrap.reason.code', string='Reason Code',
+        states={'done': [('readonly', True)]})
+    scrap_location_id = fields.Many2one(
+        'stock.location', 'Scrap Location',
+        required=True, readonly=True)
```
Comment (rousseldenis) [inline scrap_reason_code/models/stock_scrap.py]: If you want to modify just the readonly, use that attribute only.

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

**Comment by bodedra** [comment]
@rousseldenis Comment attended! Would you please review it again.

**Review by rousseldenis — APPROVED** [review]

**Comment by bodedra** [comment]
@OCA/logistics-maintainers are we good to merge this?

**Comment by bodedra** [comment]
Ping @dreispt

**Comment by rousseldenis** [comment]
@bodedra Do you want to keep the 3 commits or do you plan to squash them ?

**Comment by bodedra** [comment]
@rousseldenis Let me squash them into 1.

**Pull request merged**